### PR TITLE
Refactor how options are registered

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
Refactor how options are registered. Replaces the various ways of registering options by these four types:
- Cli: 
  - Can be chosen independent of libraries used. 
  - Never stored
  - Can be accessed through most available DafnyOptions objects
  - Examples: --cores, --verification-time-limit, --warn-as-errors
- Global: 
  - Value chosen for root project imposes constraints on this option for all consumed projects. The constraint is configured through a separate function.
  - Stored in .doo 
  - Can be accessed through most available DafnyOptions objects
  - Examples: --unicode-char, --enforce-determinism
- Module: 
  - Can be configured either through the CLI, or with an attribute at the module level
  - Stored in the doo if configured through CLI
  - Should be accessed through the instance field `ModuleDecl.Options`
  - Examples: --function-syntax, --manual-triggers, --warn-redundant-assumptions
- Translation: 
  - Can be different per module in program. 
  - Stored in .dtr 
  - Currently accessed through `program.Compilation.AlreadyTranslatedRecord.OptionsByModule`
  - Examples: --outer-module, --go-module-name

### Follow-up
- Load read translation records into `ModuleDecl.Options` and replace usages of `CompilationData.AlreadyTranslatedRecord` with `ModuleDecl.Options`, 
- Decide what to do with:
```c#
OptionRegistry.RegisterOption(OptimizeErasableDatatypeWrapper, OptionScope.Cli); // TODO needs translation record registration
OptionRegistry.RegisterOption(AddCompileSuffix, OptionScope.Cli);  // TODO needs translation record registration
```

### How has this been tested?
It's a refactoring so no additional tests are needed.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Options/OptionRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  // public static void RegisterNoChecksNeeded(Option option, bool semantic) {
+  //   if (semantic) {
+  //     RegisterLibraryCheck(option, OptionCompatibility.NoOpOptionCheck);
+  //   } else {
+  //     if (OptionChecks.ContainsKey(option)) {
+  //       throw new ArgumentException($"Option already registered as needing a library check: {option.Name}");
+  //     }
+  //     NoChecksNeeded.Add(option);
+  //   }
+  // }
```
Comment (alex-chew) [inline Source/DafnyCore/Options/OptionRegistry.cs:78]: nit: unused
```suggestion
```

File: Source/DafnyCore/DooFile.cs [hunk 2]
```diff
@@ -216,7 +215,8 @@ public class ManifestData {
       result.Options.OptionArguments[option] = libraryValue;
       result.ApplyBinding(option);
       var prefix = $"cannot load {options.GetPrintPath(libraryFile.LocalPath)}";
-      success = success && check(reporter, origin, prefix, option, localValue, libraryValue);
+      var checkpasses = OptionRegistry.GlobalCheck(option)?.Invoke(reporter, origin, prefix, option, localValue, libraryValue) ?? true;
```
Comment (alex-chew) [inline Source/DafnyCore/DooFile.cs:218]: The CI warns that `libraryValue` is possibly null, is that intended?

File: Source/DafnyCore/Options/OptionRegistry.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Options/OptionRegistry.cs:78]: Woops! Sorry

File: Source/DafnyCore/DooFile.cs (hunk 2 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/DooFile.cs:218]: Fixed

**Review by alex-chew — APPROVED** [review]
LGTM

**Pull request merged**